\documentclass[11pt]{article} 
\usepackage{amssymb, amsmath, amsthm}
\usepackage{tikz, graphicx, color, mathrsfs, rotating}
\usepackage{titlesec, lipsum}
\usepackage{fancyhdr, framed, chngcntr}

\usetikzlibrary{arrows,shapes,automata,backgrounds,decorations,petri,positioning,patterns}


\paperwidth = 8.5in
\paperheight = 11in
\textwidth = 6.5 in
\textheight = 9 in
\oddsidemargin = 0 in
\evensidemargin = 0 in
\topmargin = -.25 in
\headheight = 0.0 in
\headsep = .25 in
\footskip = .25in


\newtheorem*{repp@prob}{\repp@title}
\newcommand{\newreppprob}[2]{
\newenvironment{repp#1}[1]{
 \def\repp@title{#2 {##1}}
 \begin{repp@prob}}
 {\end{repp@prob}}}
\makeatother
\newreppprob{prob}{Problem}

\newtheorem*{repp@thm}{\repp@title}
\newcommand{\newreppthm}[2]{
\newenvironment{repp#1}[1]{
 \def\repp@title{#2 {##1}}
 \begin{repp@thm}}
 {\end{repp@thm}}}
\makeatother
\newreppthm{thm}{Theorem}


% -----------------------------------------------------------------------------
%             Macros for the course
% -----------------------------------------------------------------------------
\newcommand{\TS}{\mathcal{T}} % symbol for a topological space
\newcommand{\BS}{\mathcal{B}} % symbol for a basis
\newcommand{\R}{\mathbb{R}} % symbol for real numbers
\newcommand{\Z}{\mathbb{Z}} % symbol for integers
\newcommand{\Q}{\mathbb{Q}} % symbol for rational numbers
\newcommand{\PS}{\mathscr{P}} % symbol for power set
\newcommand{\E}{\mathbf{E}} % symbol for real numbers with Euclidean topology
\newcommand{\F}{\mathbf{F}^1} % symbol for real numbers with finite-complement topology
\renewcommand{\H}{\mathbf{H}^1} % symbol for real numbers with half-open topology
\renewcommand{\S}{\mathcal{S}} % basis topology
\newcommand{\B}{\mathbf{B}} % symbol for ball
\newcommand{\Sp}{\mathbf{S}} % symbol for sphere
\renewcommand{\int}{\operatorname{int}} % symbol for interior
\newcommand{\bnd}{\partial} % symbol for boundary
\newcommand{\homeo}{\approx} % symbol for homeomorphic


\begin{document} 


% -----------------------------------------------------------------------------
%             Start here
% -----------------------------------------------------------------------------

{\large
\noindent School of Computing %% replace "date" by the date on which the assignment was made
\hfill Chansu Park %% replace "Name" by your name

\vspace{.1in}

\noindent \today \hfill 20173245}

\vspace{.25in}

% -----------------------------------------------------------------------------
%             Erase or rearrange the options below, as necessary
% -----------------------------------------------------------------------------
\Large{
\begin{center}
\textbf{CS454}

Fall 2018, Coursework \#2
\end{center}
}

\large

% -----------------------------------------------------------------------------
%             Template for typing up an Exercise
% -----------------------------------------------------------------------------
Here is the current record of my implementation:
\begin{figure}[htb]
	\begin{center}
		\includegraphics[width=0.8\linewidth]{record.png}
	\end{center}
	\caption{A sample result of the implementation.}
\end{figure}

\section{Policies of this project} \label{sec:1}
\subsection{How to make and execute} \label{ssec:1.1}
This coursework has been built on \texttt{g++ 7.3.0} with \texttt{Ubuntu 18.04}. The language is \texttt{C++}. \\
There are three source codes: \texttt{City.cpp}, \texttt{tsp-solver.cpp} and \texttt{tsp-main.cpp} and the folder \texttt{tests} that contains some selected test cases. \\
You can find \texttt{Makefile} having three commands: \texttt{make all, make test, make clean}. \\
My implementation has 4 options which are mandatory:
\begin{itemize}
	\item[\texttt{-p}] Decides how big the population is.
	\item[\texttt{-f}] Decides the number of the fitness evaluation. In my implementation, it equals to the number of generation.
	\item[\texttt{-k}] Decides how many genes will be kept without crossover. Should be smaller that the population.
	\item[\texttt{-m}] Decides how many mutation operator will be applied for non-best genes. In reality, the mutation will be reduced as the cumulated number of generation grows.
\end{itemize}
To parse the command line, I didn't use such headers like \texttt{unistd.h}, since it depends on the operating system. Instead I parsed manually. \\
You can find the sample execution command using \texttt{make test}. The name of the executable is \texttt{tsp-solver}. It accepts the file name (relative location) first and accept 4 following options, each with followed positive integer. Note that this command sends standard output to the file \texttt{output.txt}. 

\subsection{Parsing Input} \label{ssec:1.2}
Since the announcement said that we have to stick on the format that the sample input \texttt{fl11849.txt} has, I followed that one. \\
It has 6 lines of metadata: Name of the dataset, comment, type, dimension, edge weight type, and node coordinate section. \\
Among them, I took a number at the line 4 (as \texttt{DIMENSION : number}) and used it as a size of the cities. \\
Following lines are the information of coordinates. Each line has one integer as an index of the city and two real numbers as an X/Y-coordinate. \\
So I ignored the first column and took the second and third number as coordinates. It doesn't matter whether or not it is represented by scientific notation. \\
I ignored ``EOF'' sign at the last line of the testcases since reading \texttt{<number>} lines is sufficient.

\section{Implementation} \label{sec:2}

I will only concentrate on \texttt{tsp-solver.cpp}. \texttt{City.cpp} has only helper functions about how to initialize a 2-dimensional point and computing distance between two cities; \texttt{tsp-main.cpp} just accepts the input and get a result from \texttt{tsp-solver.cpp}.\\
The main algorithm I used is Genetic Algorithm. I prepared such a size of pool, keep some best solutions with other optimization, and crossover the others using the best genes.

\subsection{Initializing seed} \label{ssec:2.1}
I first constructed Minimum Spanning Tree (MST) of given complete graph. According to the Chapter 35.2 of \cite{CLRS09}, minimum spanning tree of graph with Euclidean metric induces a tour for Traveling Salesman Person (TSP) problem with 2-approximation. It is because Euclidean metric satisfies the triangle inequality. The method is intuitive:
\begin{itemize}
	\item[1.] Select a root vertex.
	\item[2.] Compute an MSP using efficient algorithms and let the tree be $T$. (I used Prim's algorithm)
	\item[3.] Perform a preorder tree walk on $T$ starting from a root vertex. It generates an order $H$.
	\item[4.] Return $H$, which is indeed an Hamiltonian cycle.
\end{itemize}
Thus, when I initialize genes, I first constructed an MST and copied them by \texttt{<population>} times.
Since the MST is needed, I included some more data in my class \texttt{Genome}. Here are containers in \texttt{Genome}, the genes of my implementation:
\begin{itemize}
	\item[\texttt{genome}] The container of the index of cities. TSP-tour of this instance will just follow the order in this container.
	\item[\texttt{parent}] Parent information of the minimum spanning tree.
	\item[\texttt{children}] Children information of the minimum spanning tree.
	\item[\texttt{key}] Helper array to construct the minimum spanning tree.
	\item[\texttt{minSpanTreeElmt}] The array for constructing minimum spanning tree.
	\item[\texttt{tspLength}] Contains the length of the given tour.
\end{itemize}
Now the initial pool has been constructed.

\subsection{Improving the pool using GA} \label{ssec:2.2}
For each generation, my algorithm follow a simple loop:
\begin{itemize}
	\item[$\bullet$] Compute the length of tour of each genes in the pool and sort it by the length.
	\item[$\bullet$] Among $n$ genes, divide them into three classes: the best one $=:A$, other best $k-1$ genes $=:B$, the others $=:C$.
	\item[$\bullet$] Partially optimize genes in $B$.
	\item[$\bullet$] Crossover random genes in $A \cup B$ to replace $C$ by the crossover result.
	\item[$\bullet$] Mutate and partially optimize genes in $C$.
\end{itemize}
Note that I also do some optimization not only for the crossover results but also for the other genes in $B$.

\subsubsection{Partially-Matched Crossover} \label{sssec:2.2.1}
In \cite{GL85}, authors introduce a genetic operator called Partially-Matched Crossover (PMX). It accepts two genes as parents and outputs two childrens.
\begin{figure}[htb]
	\begin{center}
		\includegraphics[width=0.6\linewidth]{PMX.png}
	\end{center}
	\caption{An illustration to explain how PMX works.}
\end{figure}
The main idea is that, exchange some fixed substring and put other elements into its original location. If the element already exists, then select the alternative by predefined mapping. \\
In the above illustration, 1, 6, 8 in the left side will be substituted by 4, 5, 6, respectively, and vice versa. Because of that, when the left side tries to put 8, it first maps 8 to 6, and since it still exists, it finally maps 6 to 5 and place that.
Note that, if we put a sequence where every element uniquely exists in both parent, then the resulting sequences has every element uniquely. This operator works a little bit better than the order crossover operator introduced in the class.

\subsubsection{Mutation using Double-Bridge move} \label{sssec:2.2.2}
\begin{figure}[htb]
	\begin{center}
		\includegraphics[width=0.6\linewidth]{DBMove.png}
	\end{center}
	\caption{An illustration to explain how Double-Bridge Move works.}
\end{figure}
To mutate the order, the easiest way is flipping a subsequence of the tour so that only two entrance and two exit changes. But it cannot introduce such a big mutation. \\
Double-Bridge Move (\cite{MOF91}) introduces a new movement: cross 4 substrings. \\
Suppose that there is a tour like $\{0, ..., A-1, A, ..., B-1, B, ..., C-1, C, ..., end\}$. Then this operator cut 4 edges $\{(A-1, A), (B-1, B), (C-1, C), (end, 0)\}$ and establish new 4 edges $\{(A-1, C), (end, B), (C-1, A), (B-1, 0)\}$. 

\subsubsection{Partial optimization using 2-Opt} \label{sssec:2.2.3}
\begin{figure}[htb]
	\begin{center}
		\includegraphics[width=0.6\linewidth]{Opt2.png}
	\end{center}
	\caption{An illustration how 2-Opt performs its local optimization.}
\end{figure}
For crossovered genes and for non-best genes in the keeping list, I applied some partial optimization called 2-Opt. It is because the initial seeds are already too well-formed by MST. Performing only crossover and mutation will just increase their length in high probability so that they will just be deleted on the near future generation. \\
I performed an exhaustive search for every non-adjacent pairs of edges and applied 2-Opt operation up to the restricted mutation count. \\
To examine how many twoOpt operator will be applied to a single instance from MST, I measured the count using \texttt{rl5934.tsp} instance. It used this operator about 2000 times.

\section{Experimental results} \label{sec:3}

With the command \texttt{./tsp-solver ./tests/rl11849.tsp -p 40 -f 1500 -k 10 -m 25 > output.txt \&}, I obtained the result up to 450 generations within 21 hours. It achieved 10 generations within first 5 minutes, but the speed has been decreased since 2-Opt takes more and more when the generation cumulates. (The order of searching pair of edges starts from the leftmost ones and goes to the rightmost ones. \\
Here are some intermediate results:

\begin{align*}
	\text{Generation 1}:& 1344568.40 \\
	\text{Generation 5}:& 1334594.17 \\
	\text{Generation 20}:& 1294286.18 \\
	\text{Generation 50}:& 1235762.65 \\
	\text{Generation 100}:& 1152052.33 \\
	\text{Generation 200}:& 1006862.84 \\
\end{align*}
And I also noticed that the length didn't decrease from Generation 193. It seems that 2-Opt optimization for solutions reached a limit so that no optimization will occur for the best $k$ keeping parents, and the partial optimization for children may not exceed the parents. \\
For the smaller instances like \texttt{pr76.tsp}, some noticible breakthroughs has occurred at once per 100-200 generations. If I started a little bit earlier, then I might be able to performing more intense optimization for children will help them to break parent pool and test them. Or maybe some breakthrough will be present after some more generation.
I inserted \texttt{debug-rl11849.txt} for the instance for the leaderboard and \texttt{debug-pr76.txt} for the \texttt{pr76.tsp} instance in the folder \texttt{./tsp}.

\section{To debug and find intermediate results}
In \texttt{tsp-main.cpp} and \texttt{tsp-solver.cpp}, there is a define statement: \texttt{\#define DEBUG ...}. If you set this as 1, it prints every 5th intermediate tours and length for every generation, so you can observe them for large dataset. It also picks its metadata.

\begin{thebibliography}{9}
	\bibitem{CLRS09}
	Cormen, T. H., Leiserson, C. E., Rivest, R. L., \& Stein, C. (2009). Introduction to algorithms. MIT press.
	\bibitem{GL85}
	Goldberg, D. E., \& Lingle, R. (1985, July). Alleles, loci, and the traveling salesman problem. In Proceedings of an international conference on genetic algorithms and their applications (Vol. 154, pp. 154-159). Lawrence Erlbaum, Hillsdale, NJ.
	\bibitem{MOF91}
	Martin, O., Otto, S. W., \& Felten, E. W. (1991). Large-step Markov chains for the traveling salesman problem.
\end{thebibliography}
%\begin{proof}[Solution.]
%% write your solution here
%\end{proof}


% -----------------------------------------------------------------------------
%             Template for typing up a Theorem
% -----------------------------------------------------------------------------
%\begin{reppthm}{42} %% replace "42" by the relevant Theorem number
%% restate the Theorem here
%\end{reppthm}

%\begin{proof}
%% write your proof here
%\end{proof}


% -----------------------------------------------------------------------------
%             End here
% -----------------------------------------------------------------------------

\end{document}